src/expr/src/expr/expr_array_concat.rs

```diff
@@ -0,0 +1,488 @@
+                panic!("the rhs must be compatible to append to lhs");
+            }
+        }
+    }
+
+    /// Prepends an array as the front element of an array of array.
+    /// Note the behavior is slightly different from PG.
+    ///
+    /// Examples:
+    /// - `select array_cat(array[233], array[array[66]]);` => `[[233], [66]]`
+    /// - `select array_cat(null::int[], array[array[66]]);` => `[[66]]` ignore NULL, same as PG
+    /// - `select array_cat(array[233], null::int[][]);` => `null` different from PG
+    /// - `select array_cat(null::int[], null::int[][]);` => `null` same as PG
+    fn prepend_array(left: DatumRef, right: DatumRef) -> Datum {
+        match (left, right) {
+            (_, None) => None,
+            (None, right @ Some(ScalarRefImpl::List(_))) => {
+                right.map(ScalarRefImpl::into_scalar_impl)
+            }
+            (left, Some(ScalarRefImpl::List(right))) => Some(
+                ListValue::new(
+                    std::iter::once(left)
+                        .chain(right.values_ref().into_iter())
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
+                        .collect(),
+                )
+                .into(),
+            ),
+            _ => {
+                panic!("the lhs must be compatible to prepend to rhs");
+            }
+        }
+    }
+
+    /// Prepends a value as the front element of an array.
+    /// The behavior is the same as PG.
+    ///
+    /// Examples:
+    /// - `select array_prepend(123, array[66]);` => `[123, 66]`
+    /// - `select array_prepend(null::int, array[66]);` => `[null, 66]`
+    /// - `select array_prepend(233, null::int[]);` => `[233]`
+    /// - `select array_prepend(null::int, null::int[]);` => `[null]`
+    fn prepend_value(left: DatumRef, right: DatumRef) -> Datum {
+        match (left, right) {
+            (left, None) => {
+                Some(ListValue::new(vec![left.map(ScalarRefImpl::into_scalar_impl)]).into())
+            }
+            (left, Some(ScalarRefImpl::List(right))) => Some(
+                ListValue::new(
+                    std::iter::once(left)
+                        .chain(right.values_ref().into_iter())
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
+                        .collect(),
+                )
+                .into(),
+            ),
+            _ => {
+                panic!("the lhs must be compatible to prepend to rhs");
+            }
+        }
+    }
+
+    fn dispatch(&self, left: DatumRef, right: DatumRef) -> Datum {
+        match self.op {
+            Operation::ConcatArray => Self::concat_array(left, right),
+            Operation::AppendArray => Self::append_array(left, right),
+            Operation::AppendValue => Self::append_value(left, right),
+            Operation::PrependArray => Self::prepend_array(left, right),
+            Operation::PrependValue => Self::prepend_value(left, right),
+        }
+    }
+}
```

Comment:
Use `op: fn(DatumRef, DatumRef) -> Datum`, then `dispatch` is eliminated 🤪

BTW, is there any difference between using a function pointer vs using separate expr types? cc @skyzh

Reply:
There might be some lifetime issues, but I think there's no much difference.